### ISSUE TITLE
controlsfx, CheckComboBox

### ISSUE THREAD

**Issue by lsh-0** [cljfx/cljfx#156]
This is not a `cljfx` issue, but I'm trying to wrap the [CheckComboBox](https://controlsfx.github.io/javadoc/11.1.1/org.controlsfx.controls/org/controlsfx/control/CheckComboBox.html#checkModelProperty) control for cljfx and was wondering if you had some insight?

This is what I've pasted together so far from cljfx examples and other controls:

```clojure
(defn- observable-list [x]
  (cond
    (instance? ObservableList x) x
    (instance? Collection x) (FXCollections/observableArrayList ^Collection x)
    :else (coerce/fail ObservableList x)))

(def with-ext-props
  (fx/make-ext-with-props
   {:on-selected-items-changed (prop/make
                                 (mutator/list-change-listener
                                  #(.getCheckedItems ^IndexedCheckModel (.getCheckModel ^CheckComboBox %)))
                                 lifecycle/list-change-listener)}))

(def props
  (merge
   fx.control/props
   (composite/props CheckComboBox
                    :title [:setter lifecycle/scalar]
                    :items [:getter lifecycle/scalar :coerce observable-list]
                    :show-checked-count [:setter lifecycle/scalar])))

(def lifecycle
  (composite/describe CheckComboBox
                      :ctor [:items]
                      :props props))
```

and it mostly works using something like this:

```clojure
  {:fx/type controlsfx.check-combo-box/with-ext-props
   :props {:on-selected-items-changed println}
   :desc {:fx/type controlsfx.check-combo-box/lifecycle
          :title "some title"
          :items ["foo" "bar" "baz"]
          :show-checked-count true}}
```

however, re-rendering that with an updated `:items` fails with the error:

> java.lang.IllegalArgumentException: No implementation of method: :replace! of protocol: #'cljfx.mutator/Mutator found for class: clojure.lang.Keyword

presumably because the `items` can only be set during the construction of `CheckComboBox`. 

If that is the case then the control needs to be replaced when the items change but I haven't figured that out.

I'd also like to modify the [selected check boxes](https://controlsfx.github.io/javadoc/11.1.1/org.controlsfx.controls/org/controlsfx/control/IndexedCheckModel.html) but how do I attach a custom setter to the control description like `:checked` or `:selected`?

I can't quite make heads or tails of where or what I should be looking at in the code. Any help?

**Comment by dsbw** [comment]
The tl;dr: If you build your views using contexts, you can re-render a component which only exposes its items through an `ObservableList` (where most use `ObjectProperty`).

----
OK, I struggled through this one and took the following steps:

1. I noted that ControlsFX's `CheckComboBox`'s `items` property is an ObservableList _but it's not an ObjectProperty_. That seems to distinguish it from all the dynamic instance examples I can find, and in fact it seems most of the JavaFX components wrap their OLs in OPs.
2. I did note that `SplitPane` has a plain old OL items property so it could maybe serve as an analog to the `CheckComboBox`. (I don't want to mess with ControlsFX at the moment.) Sure enough, if I take the first atom example (where the title changes based on entry into the text field), I could not make that work directly for `SplitPane`. (`TitledPane` wraps its content in an ObjectProperty.)
3. The `items` property has to be a list, it can't be a function, so we can rule out that, unfortunately.
4. So I went to e09, the to-do list and basically ended up in the same place. We can dynamically change out contents for `ObjectProperty` but that won't force the re-rendering of a component based on `ObservableList`.
5. Next up is e16, but we don't have a factory method, alas. 
6. Creating a new version of the component that had its items wrapped in an OP did occur to me, but I didn't want to give up without looking at contexts, ext-props and lifecycles.
7. So I tried context. I didn't do this first thing because (in theory) the whole context approach is optional. But if you're using context (as I am), it works in such a way that you might never notice a difference between an OL and an OP-type approach.

Here's an example of add items to a `SplitPane`, where `items` is defined as an `ObservableList` with no `ObjectProperty` wraparound:
```clojure
(ns ezclj.utils.context
  (:require [cljfx.api :as fx]
            [clojure.core.cache :as cache]))

(def *context
  (atom (fx/create-context {:items [{:fx/type :label :text "Starter"}]} cache/lru-cache-factory)))

(defn event-handler [{:keys [fx/event event/type fx/context]}]
  (if (= type :add)
    (let [items (fx/sub-val context :items)]
      {:context (fx/swap-context context assoc :items (conj items {:fx/type :label :text "new label"}))})))

(defn root [context]
  (let [items (fx/sub-val context :items)]
    {:fx/type :stage
     :showing true
     :scene   {:fx/type :scene
               :root    {:fx/type          :split-pane
                         :items            items
                         :on-mouse-clicked {:event/type :add}}}}))

(def app
  (fx/create-app
    *context
    :event-handler event-handler
    :desc-fn (fn [context] (root context))))
```
You can see that the clicking causes new panes to be added. I suspect that if you build your checkboxes with a `sub-val` item list, you'll get the same results. I'm sure this isn't the only answer, but I think it will work.

**Comment by vlaaad** [comment]
Hi! `composite/props` does not support `:getter` keyword. It probably should throw exception on receiving keyword it does not recognize... What it does support is:
- `:setter`
- `:list`
- `:map`
- `:property-change-listener`

**Comment by vlaaad** [comment]
I also wrote [infer-props](https://github.com/cljfx/cljfx/blob/master/dev/user.clj) helper function that could be used as a starting point for building props, though its results might require a lot of post-processing. Here is its output for CheckComboBox:
```clj
(import '[org.controlsfx.control CheckComboBox])

(infer-props CheckComboBox)
=> (:check-model [:setter org.controlsfx.control.IndexedCheckModel :default #object[org.controlsfx.control.CheckComboBox$CheckComboBoxBitSetCheckModel 0x690c3b1f "org.controlsfx.control.CheckComboBox$CheckComboBoxBitSetCheckModel@690c3b1f"]]
    :converter [:setter javafx.util.StringConverter]
    :items [:list #clojure.reflect.Method{:name getItems
                                          :return-type javafx.collections.ObservableList
                                          :declaring-class org.controlsfx.control.CheckComboBox
                                          :parameter-types []
                                          :exception-types []
                                          :flags #{:public}}]
    :show-checked-count [:setter lifecycle/scalar :default false]
    :title [:setter lifecycle/scalar])
```

**Comment by vlaaad** [comment]
After cleaning this up I managed to create a working CheckComboBox lifecycle:
```clj
(require '[cljfx.composite :as composite]
         '[cljfx.lifecycle :as lifecycle]
         '[cljfx.mutator :as mutator]
         '[cljfx.coerce :as coerce]
         '[cljfx.api :as fx])

(def props
  (merge cljfx.fx.control/props
         (composite/props CheckComboBox
           :items [:list lifecycle/scalar]
           :converter [:setter lifecycle/scalar :coerce coerce/string-converter]
           :show-checked-count [:setter lifecycle/scalar :default false]
           :title [:setter lifecycle/scalar]
           :check-model [:setter lifecycle/scalar]
           :on-checked-items-changed [(mutator/list-change-listener
                                        #(.getCheckedItems (.getCheckModel ^CheckComboBox %)))
                                      lifecycle/list-change-listener])))

(def lifecycle
  (composite/describe CheckComboBox
    :ctor []
    :props props))

(fx/on-fx-thread
  (fx/create-component
    {:fx/type :stage
     :showing true
     :scene {:fx/type :scene
             :root {:fx/type :v-box
                    :children [{:fx/type lifecycle
                                :items [1 2 3 4]
                                :on-checked-items-changed tap>}]}}}))
```

**Comment by vlaaad** [comment]
So you almost got everything right in the first place, you just needed to define `:items` with `[:list lifecycle/scalar]`. Sorry that cljfx doesn't help enough with writing custom composite lifecycles. Did it help?

**Comment by lsh-0** [comment]
This looks very helpful, thank you @vlaaad . I'll sit down tonight and have a proper go at it.

**Comment by lsh-0** [comment]
This works well, thanks for your time and attention @vlaaad.
